shared/js/dataDriver.js

```diff
@@ -0,0 +1,283 @@
+  "use strict";
+
+  let { actions } = loop.shared;
+
+  /**
+   * Generate an id based on a time. Declared outside of the class to use
+   * locally scoped variables.
+   * https://gist.github.com/mikelehen/3596a30bd69384624c11
+   */
+  let makeId = function() {
+    let CHARS = "0123456789ABCDEFGHIJKLMNOPQRSTUV" +
+                "WXYZ_abcdefghijklmnopqrstuvwxyz~",
+        lastRandChars = [],
+        lastTime;
+
+    return time => {
+      let duplicateTime = time === lastTime;
+      lastTime = time;
+
+      let timeStampChars = [];
+      for (let i = 7; i >= 0; i--) {
+        timeStampChars.unshift(CHARS.charAt(time % 64));
+        time = Math.floor(time / 64);
+      }
+      if (time !== 0) {
+        throw new Error("We should have converted the entire timestamp.");
+      }
+
+      let id = timeStampChars.join("");
+      if (duplicateTime) {
+        let i;
+        for (i = 11; i >= 0 && lastRandChars[i] === 63; i--) {
+          lastRandChars[i] = 0;
+        }
+        lastRandChars[i]++;
+      }
+      else {
+        for (let i = 0; i < 12; i++) {
+          lastRandChars[i] = Math.floor(Math.random() * 64);
+        }
+      }
+      for (let i = 0; i < 12; i++) {
+        id += CHARS.charAt(lastRandChars[i]);
+      }
+      if (id.length !== 20) {
+        throw new Error("Length should be 20.");
+      }
+
+      return id;
+    };
+  }();
+
+  return class DataDriver {
+    /** ************ **
+     *** DataDriver ***
+     ** ************ **/
+
+    /**
+     * Create a DataDriver and initialize server connections.
+     */
+    constructor(options) {
+      if (!options.dispatcher) {
+        throw new Error("Missing option dispatcher");
+      }
+
+      this._dispatcher = options.dispatcher;
+      this._dispatcher.register(this, [
+        "fetchServerData",
+        "setupWindowData"
+      ]);
+    }
+
+    /**
+     * Send a text chat message by storing in the database.
+     */
+    sendTextChatMessage(message) {
+      this.update("chat", this.makeId(), message);
+    }
+
+    /** **************** **
+     *** Action Handler ***
+     ** **************** **/
+
+    /**
+     * Handle fetchServerData action by connecting to the room.
+     */
+    fetchServerData({ token }) {
+      this._connectToRoom(token);
+    }
+
+    /**
+     * Handle setupWindowData action by connecting to the room.
+     */
+    setupWindowData({ roomToken }) {
+      this._connectToRoom(roomToken);
+    }
+
+    /** *************** **
+     *** EventListener ***
+     ** *************** **/
+
+    /**
+     * Handle various events from addEventListener.
+     */
+    handleEvent(event) {
+      switch (event.type) {
+        case "open":
+          this._dispatcher.dispatch(new actions.DataChannelsAvailable({
+            available: true
+          }));
+          break;
+
+        case "put": {
+          // Ignore initial put from connecting.
+          let { data, path } = JSON.parse(event.data);
```

Comment:
Wrapped. I suppose this parse could fail if Firebase changes how it sends updates via streaming events.